[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62419
* #62418
* #62417
* #62228
* #62191
* #61792

Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

kimishpatel committed:
[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler

Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile. This bring cpp API for
mobile profiling on part with Torchscript.
This is done via:
1. Utilizating debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState
3. Creating new RAII stype profiler, KinetoEdgeCPUProfiler, which can be
used in surrounding scope of model execution. This will write chrome
trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile. This bring cpp API for
mobile profiling on part with Torchscript.
This is done via:
1. Utilizating debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState
3. Creating new RAII stype profiler, KinetoEdgeCPUProfiler, which can be
used in surrounding scope of model execution. This will write chrome
trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile. This bring cpp API for
mobile profiling on part with Torchscript.
This is done via:
1. Utilizating debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState
3. Creating new RAII stype profiler, KinetoEdgeCPUProfiler, which can be
used in surrounding scope of model execution. This will write chrome
trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile. This bring cpp API for
mobile profiling on part with Torchscript.
This is done via:
1. Utilizating debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState
3. Creating new RAII stype profiler, KinetoEdgeCPUProfiler, which can be
used in surrounding scope of model execution. This will write chrome
trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel commented:
@iseeyuan, @raziel in this PR we are building cpu only kineto with lite interpreter. Is this ok for oss binary size or do we want to make it optional?

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile. This bring cpp API for
mobile profiling on part with Torchscript.
This is done via:
1. Utilizating debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState
3. Creating new RAII stype profiler, KinetoEdgeCPUProfiler, which can be
used in surrounding scope of model execution. This will write chrome
trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile. This bring cpp API for
mobile profiling on part with Torchscript.
This is done via:
1. Utilizating debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState
3. Creating new RAII stype profiler, KinetoEdgeCPUProfiler, which can be
used in surrounding scope of model execution. This will write chrome
trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

raziel commented on torch/csrc/jit/mobile/profiler_edge.cpp:
```diff
@@ -0,0 +1,48 @@
+    const bool with_stack,
+    const bool with_flops,
+    const bool with_module_hierarchy) :
+  m_(m), trace_file_name_(fname) {
```
The dependency to m I don't quite understand well.
In the docs you said that m_ exists to make an implicit dependency on the Module explicit, but can you clarify what's the implicit dependency coming from?

kimishpatel replied:
OK I think these comments went unnoticed due to the code being outdated. So implicit dependency is coming from the fact that post processing callback lambda captures m, so we dont really need to explicitly capture it as part of EdgeProfiler.

raziel reviewed (changes requested):
Thanks

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel commented on cmake/Dependencies.cmake:
```diff
@@ -1565,6 +1565,11 @@ endif()
 
+# we need to build all targets to be linked with PIC
+if(USE_KINETO AND INTERN_BUILD_MOBILE AND USE_LITE_INTERPRETER_PROFILER)
+  set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
```
I was getting relocation error during linking and I think it should have been already compiling with `-fPIC` for some reason without setting this flag it was not doing that.

kimishpatel commented on cmake/Dependencies.cmake:
```diff
@@ -1873,11 +1878,17 @@ list(APPEND Caffe2_DEPENDENCY_LIBS fmt::fmt-header-only)
   set(USE_KINETO OFF)
 endif()
 
+if(USE_KINETO AND INTERN_BUILD_MOBILE AND USE_LITE_INTERPRETER_PROFILER AND (USE_CUDA OR USE_ROCM))
```
Unfortunately, not right now. But I think we should be able to just add `USE_CPU_ONLY`. And we can set this when `NOT USE_CUDA AND NOT USE_ROCM`

kimishpatel commented on torch/csrc/jit/mobile/debug_info.cpp:
```diff
@@ -82,20 +85,21 @@ std::pair<std::string, std::string> getStackTraceWithModuleHierarchy(
   std::string module_info =
       root_scope_string + "(" + top_module_type_name + ")";
-  std::string caller_fn_name = "FunctionName_UNKNOWN";
+  std::string caller_fn_name = "<forward>";
```
For hardcoded to <forward>: This is because we use debug handles for symbolication. So debug handle correspond to some node in a graph like:
```
graph(%self, %in):
%x = getAttr...
%1 = aten::add
```
The graph may correspond to some function name `my_forward`. And in this case debug handle for %1 does not know that it is for `my_forward`. So we dont have access to that info.

I think this can be fixed but maybe in a separate PR.

> and then immediately appended (mainly, why weren't we appending it before --are we potentially doing it twice, given that above we also added module_info.append("::").append(prev_function_name))?

Good observation, but we are not doing that twice because `caller_fn_name` is passed to `getStackTraceWithModuleHierarchy` but in that function when we do `module_info.append("::").append(prev_function_name);`, that happens after `prev_function_name` updated from the `InlinedCallStack` corresponding to debug handle. That will be different from `caller_fn_name` because as you can see in, https://github.com/pytorch/pytorch/pull/62419/files#diff-fbfe49a9e94202548c66b89039d024a473f8fbf2d0c9f42a54cfe7a1d08499dfL52-L58, `prev_function_name` gets updated.

kimishpatel replied:
Also I will update function name from `<forward>` to `unknown`

kimishpatel commented on torch/csrc/jit/mobile/module.cpp:
```diff
@@ -1,5 +1,6 @@
 #include <torch/csrc/jit/mobile/module.h>
 
+#include <bits/stdint-intn.h>
```
This is my vim and lsp setup messing things up. Thanks for pointing this out. WIll fix.

kimishpatel commented on torch/csrc/jit/mobile/profiler_edge.h:
```diff
@@ -0,0 +1,61 @@
+class TORCH_API KinetoEdgeCPUProfiler {
+ public:
+  // This profiler only profiler KINETO events
+  // No GPU_FALLBACK or NVTX
+  KinetoEdgeCPUProfiler(
```
Nice! Thanks for this. Incorporating this.

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

raziel commented on torch/csrc/autograd/profiler_kineto.h:
```diff
@@ -334,6 +338,28 @@ TORCH_API void enableProfiler(
+ * KinetoEvents.
+ * enableProfilerWithEventPostProcess enables profiler to capture
+ * specified activities, with specified RecordFunction scope, if any.
+ * Additinally it takes functor that does in-place post processing of
```
```suggestion
 * Additionally, it takes a functor that does in-place post processing of
```

raziel reviewed (approved):
Thanks!

There's just a typo fix, and still a couple of comments left unanswered that I'd like to understand more. Particularly this one:

> Why do we not instrument the other instructions? Particularly OPN and I believe INTERFACE_CALL which will be used to call upon delegates and other custom classes?

kimishpatel commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -81,7 +83,7 @@ bool InterpreterState::run(Stack& stack) {
             enableRecordFunction(true);
           }
-          RECORD_USER_SCOPE_WITH_INPUTS(code_->op_names_[inst.X].name, stack);
+          RECORD_EDGE_SCOPE_WITH_DEBUG_HANDLE_AND_INPUTS(code_->op_names_[inst.X].name, debug_handle, stack);
```
This also went unnoticed for the same reason. I need to change my rebuttal style.

Ok this is a good callout. Particular INTERFACE_CALL and OPN. I will add those changes. Rest of the prim instructions and native instructions may not be worth it and may add overhead. Although if we want to capture them we can in the future?

raziel replied:
Sounds good.

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch Edge] Enable kineto profiler on mobile via EdgeKinetoProfiler"


Summary:
This diff adds support for cpu only kineto profiler on mobile. Thus
enabling chrome trace generation on mobile.

This brings the cpp API for mobile profiling on par with Torchscript.

This is done via:
1. Utilizing debug handle annotations in KinetoEvent.
2. Adding post processing capability, via callbacks, to
KinetoThreadLocalState.
3. Creating new RAII-style profiler, KinetoEdgeCPUProfiler, which can be
used as a scope of model execution that we aim to profile. This will write
a  chrome trace to the location specified in profiler constructor.

Test Plan:
MobileProfiler.ModuleHierarchy

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29993660](https://our.internmc.facebook.com/intern/diff/D29993660)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29993660).
